### ISSUE TITLE
Comparison of Union of sets with identical srepr is False

### ISSUE THREAD

**Issue by oscargus** [sympy/sympy#21943]
From #21925
```
def test_issue_13396():
    expr = -2*y*exp(-x**2 - y**2)*Abs(x)
    # The second part may be empty
    sol = Union(FiniteSet(Integer(0)), Intersection(S.Reals, FiniteSet(Pow(Add(Mul(Integer(-1),
                Pow(x, Integer(2))), zoo), Rational(1, 2)), Mul(Integer(-1),
                 Pow(Add(Mul(Integer(-1), Pow(x, Integer(2))), zoo), Rational(1, 2))))))

    # Direct comparison does not work
    assert srepr(solveset(expr, y, domain=S.Reals)) == srepr(sol)
    # Original issue was TypeError
```
The following comparison gives `True`, but comparing the result and `sol` directly gives `False`

**Comment by oscarbenjamin** [comment]
The args are in a different order:
```python
In [43]: f = lambda e: e.args[1].args[1].args[0].args[0]

In [44]: f(sol)
Out[44]: 
   2      
- x  + zoo

In [45]: f(sol2)
Out[45]: 
   2      
- x  + zoo

In [46]: f(sol).args
Out[46]: 
⎛       2⎞
⎝zoo, -x ⎠

In [47]: f(sol2).args
Out[47]: 
⎛  2     ⎞
⎝-x , zoo⎠
```
The output of `srepr` should not be the same for both:
```python
In [48]: srepr(f(sol))
Out[48]: "Add(Mul(Integer(-1), Pow(Symbol('x'), Integer(2))), zoo)"

In [49]: srepr(f(sol2))
Out[49]: "Add(Mul(Integer(-1), Pow(Symbol('x'), Integer(2))), zoo)"
```

**Comment by oscargus** [comment]
OK! Probably these are the culprit lines:
https://github.com/sympy/sympy/blob/63b8665369344ed7a63e3a9f9b41f27ed76fa809/sympy/printing/repr.py#L179-L183

While sorting makes sense in all other printing, it seems like it shouldn't be used for repr/srepr...

**Comment by oscarbenjamin** [comment]
Another problem is that `evaluate=False` is actually needed to create expressions with out of order args:
```python
In [50]: Add(x, zoo).args
Out[50]: (zoo, x)

In [51]: Add(x, zoo, evaluate=False).args
Out[51]: (x, zoo)
```
I'm not sure how to check whether `evaluate=False` is needed apart from `Add(*args).args == args`.

**Comment by oscargus** [comment]
Yes, but 
```
In [38]: Add(zoo, x).args
Out[38]: (zoo, x)
```
so as long as it prints in a way that can be used without any further sorting it shouldn't really be a problem here. To me it seems like `srepr(expr1) == srepr(expr2)` should give `expr1 == expr2`, not necessarily that the arguments are returned in the same way to what is entered.

**Comment by oscarbenjamin** [comment]
To me the important identity is (assuming imports):
```python
expr == eval(srepr(expr))
```
That is `srepr` gives a textual representation that preserves complete structural information to rebuild the expression exactly.

Currently:
```python
In [61]: expr = Add(x, zoo, evaluate=False)

In [62]: expr == eval(srepr(expr))
Out[62]: False
```

**Comment by oscargus** [comment]
Personally, I'm not so worried about the `evaluate=False` case as this is something that will in general be very tricky to change (just see all the printing issue) and find it much more relevant that the regular use case works rather than trying to bypass some of the features... However, that case should also work by removing the use of sets (if any) and any type of sorting of arguments in the repr printer I would presume.

**Comment by oscargus** [comment]
I mean, just look at this:
https://github.com/sympy/sympy/blob/63b8665369344ed7a63e3a9f9b41f27ed76fa809/sympy/core/add.py#L24-L27

Argument sorting is well built in, even for this. But printing the srepr so that it can be exactly recreated seems like a small step.

**Comment by oscarbenjamin** [comment]
I agree. The number one thing that printing should do is show when expressions are not the same. This is especially important for `srepr` because it is supposed to be a completely unambiguous printer.
